File: Core/Models/PeerBenchmark.swift
import Foundation

/// Materialised cohort statistics for a single metric. One row per
/// (cohort definition, metric) tuple. Any cohort dimension left nil means
/// "across all values for that dimension" — so a benchmark with only
/// `ageDivision` set is the cohort of all athletes in that age group
/// regardless of belt or weight.
public struct PeerBenchmark: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var beltRank: BeltRank?
    public var ageDivision: AgeGroup?
    public var weightClass: WeightCategory?
    /// Canonical metric identifier. Typically a `PhysicalMetricKind` raw value.
    public var metricKey: String
    public var mean: Double
    public var standardDeviation: Double
    public var sampleSize: Int
    public var computedAt: Date

    public init(
        id: EntityID = UUID(),
        beltRank: BeltRank? = nil,
        ageDivision: AgeGroup? = nil,
        weightClass: WeightCategory? = nil,
        metricKey: String,
        mean: Double,
        standardDeviation: Double,
        sampleSize: Int,
        computedAt: Date = Date()
    ) {
        self.id = id
        self.beltRank = beltRank
        self.ageDivision = ageDivision
        self.weightClass = weightClass
        self.metricKey = metricKey
        self.mean = mean
        self.standardDeviation = standardDeviation
        self.sampleSize = sampleSize
        self.computedAt = computedAt
    }

    /// True when this benchmark applies to the given athlete (each cohort
    /// dimension either matches or is unconstrained on the benchmark).
    public func appliesTo(_ athlete: Athlete) -> Bool {
        if let rank = beltRank, rank != athlete.currentBelt.rank { return false }
        if let age = ageDivision, age != athlete.ageGroup { return false }
        if let cat = weightClass, cat != athlete.weightClass { return false }
        return true
    }

    /// z-score = (value − mean) / σ. Returns nil when σ is zero.
    public func zScore(for value: Double) -> Double? {
        guard standardDeviation > 0 else { return nil }
        return (value - mean) / standardDeviation
    }
}

public extension Array where Element == PeerBenchmark {
    /// Best-fit benchmark for the athlete + metric — prefer the most
    /// specific cohort match (more dimensions set), fall back to
    /// progressively broader cohorts. Returns nil when no benchmark exists
    /// for the metric at all.
    func best(for athlete: Athlete, metricKey: String) -> PeerBenchmark? {
        self
            .filter { $0.metricKey == metricKey && $0.appliesTo(athlete) }
            .max(by: { lhs, rhs in lhs.specificity < rhs.specificity })
    }
}

private extension PeerBenchmark {
    /// Number of cohort dimensions explicitly set. Used to pick the most
    /// specific applicable benchmark.
    var specificity: Int {
        var n = 0
        if beltRank != nil { n += 1 }
        if ageDivision != nil { n += 1 }
        if weightClass != nil { n += 1 }
        return n
    }
}

File: flutter_app/lib/core/models/peer_benchmark.dart
import 'athlete.dart';
import 'belt.dart';
import 'entity_id.dart';
import 'tournament.dart';

/// Port of Core/Models/PeerBenchmark.swift.
///
/// Materialised cohort statistics for a single metric. One row per
/// (cohort definition, metric) tuple. Any cohort dimension left null means
/// "across all values for that dimension".
class PeerBenchmark {
  final EntityID id;
  final BeltRank? beltRank;
  final AgeGroup? ageDivision;
  final WeightCategory? weightClass;

  /// Canonical metric identifier. Typically a [PhysicalMetricKind] raw value.
  final String metricKey;
  final double mean;
  final double standardDeviation;
  final int sampleSize;
  final DateTime computedAt;

  PeerBenchmark({
    EntityID? id,
    this.beltRank,
    this.ageDivision,
    this.weightClass,
    required this.metricKey,
    required this.mean,
    required this.standardDeviation,
    required this.sampleSize,
    DateTime? computedAt,
  }) : id = id ?? newEntityId(),
       computedAt = computedAt ?? DateTime.now();

  /// True when this benchmark applies to the given athlete (each cohort
  /// dimension either matches or is unconstrained on the benchmark).
  ///
  /// NOTE: [Athlete.weightClass] (optional override stored on the athlete
  /// record) is deferred — until it lands, [WeightCategory.suggested] is used
  /// as a proxy so benchmark matching works end-to-end today.
  bool appliesTo(Athlete athlete) {
    if (beltRank != null && beltRank != athlete.currentBelt.rank) return false;
    if (ageDivision != null && ageDivision != athlete.ageGroup) return false;
    if (weightClass != null) {
      final athleteWeightClass = WeightCategory.suggested(athlete);
      if (athleteWeightClass != weightClass) return false;
    }
    return true;
  }

  /// z-score = (value − mean) / σ. Returns null when σ is zero.
  double? zScore(double value) {
    if (standardDeviation <= 0) return null;
    return (value - mean) / standardDeviation;
  }

  /// Number of cohort dimensions explicitly set (used for most-specific match).
  int get _specificity {
    var n = 0;
    if (beltRank != null) n++;
    if (ageDivision != null) n++;
    if (weightClass != null) n++;
    return n;
  }

  factory PeerBenchmark.fromJson(Map<String, dynamic> json) => PeerBenchmark(
    id: json['id'] as String,
    beltRank: json['beltRank'] != null
        ? _beltRankFromJson(json['beltRank'] as Map<String, dynamic>)
        : null,
    ageDivision: json['ageDivision'] != null
        ? AgeGroup.fromJson(json['ageDivision'] as String)
        : null,
    weightClass: json['weightClass'] != null
        ? WeightCategory.fromJson(json['weightClass'] as String)
        : null,
    metricKey: json['metricKey'] as String,
    mean: (json['mean'] as num).toDouble(),
    standardDeviation: (json['standardDeviation'] as num).toDouble(),
    sampleSize: json['sampleSize'] as int,
    computedAt: DateTime.parse(json['computedAt'] as String),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'beltRank': beltRank != null
        ? {'kind': beltRank!.kind.name, 'number': beltRank!.number}
        : null,
    'ageDivision': ageDivision?.name,
    'weightClass': weightClass?.name,
    'metricKey': metricKey,
    'mean': mean,
    'standardDeviation': standardDeviation,
    'sampleSize': sampleSize,
    'computedAt': computedAt.toIso8601String(),
  };
}

BeltRank _beltRankFromJson(Map<String, dynamic> json) => BeltRank(
  kind: BeltKind.fromJson(json['kind'] as String),
  number: json['number'] as int,
);

extension PeerBenchmarkListExt on List<PeerBenchmark> {
  /// Best-fit benchmark for the athlete + metric — prefer the most specific
  /// cohort match (more dimensions set), fall back to progressively broader
  /// cohorts. Returns null when no benchmark exists for the metric at all.
  PeerBenchmark? best(Athlete athlete, String metricKey) {
    final candidates = where(
      (b) => b.metricKey == metricKey && b.appliesTo(athlete),
    ).toList();
    if (candidates.isEmpty) return null;
    candidates.sort((a, b) => b._specificity.compareTo(a._specificity));
    return candidates.first;
  }
}
